_posts/2017-11-23-totw-109.md

```diff
@@ -0,0 +1,111 @@
+permalink: tips/109
+type: markdown
+order: "109"
+---
+
+Originally posted as totw/109 on 2016-01-14
+
+By Greg Miller (jgm@google.com)
+
+This document will explain when `const` is meaningful in function declarations,
+and when it is meaningless and best omitted. But first, let us briefly explain
+what is meant by the terms *declaration* and *definition*.
+
+Consider the following code.
+
+```c++
+void F(int);                     // 1: declaration of F(int)
+void F(const int);               // 2: re-declaration of F(int)
+void F(int) { /* ... */ }        // 3: definition of F(int)
+void F(const int) { /* ... */ }  // 4: error: re-definition of F(int)
+```
+
+The first two lines are function *declarations*. A function *declaration* tells
+the compiler the function's signature and return type. In the above example, the
+function's signature is `F(int)`. The constness of the function's parameter type
+is ignored, so both declarations are equivalent ([over.load][CppStandard]).
+
+Lines 3 and 4 from the above code are both function *definitions*. A function
+*definition* is also a declaration, but a definition also contains the
+function's body. Therefore, line 3 is a definition for a function with the
+signature `F(int)`. Similarly, line 4 is also a definition for the same
+function, `F(int)`, which will result in an error at link time. Multiple
+declarations are allowed, but only a single definition is permitted.
+
+Even though the definitions on lines 3 and 4 *declare* and *define* the same
+function, there is a difference within their function bodies due to the way they
+are declared. From the definition on line 3, the type of the function-parameter
+variable within the function will be `int` (i.e., non-const). On the other hand,
+the definition on line 4 will produce a function-parameter variable within the
+function whose type is `const int`.
+
+### Meaningful `const` in function declarations
+
+Not all `const` qualifications in function declarations are ignored. To quote
+from [over.load][CppStandard] in the C++ standard (emphasis added):
+
+> `const` type-specifiers **buried within a parameter type specification** are
+> significant and can be used to distinguish overloaded function declarations
+
+The following are examples where `const` is significant and not ignored.
+
+```c++
+void F(const int* x);                  // 1
+void F(const int& x);                  // 2
+void F(std::unique_ptr<const int> x);  // 3
+void F(int* x);                        // 4
+```
+
+In the above examples, the `x` parameter itself is never declared `const`. Each
+of the above functions accepts a parameter named `x` of a different type, thus
+forming a valid overload set. Line 1 declares a function that accepts a "pointer
+to an `int` that is `const`". Line 2 declares a function that accepts a
+"reference to an `int` that is `const`". And line 3 declares a function that
+accepts a "unique_ptr of `int`s that are `const`". All of these uses of `const`
```

Comment:
of an \`int`\ that is (?)


(The way it's phrased suggests it accepts an array of ints ...)

Reply:
I don't follow

Reply:
The original comment is plural

declares a function that accepts a "unique_ptr of `int`s that are `const`"

vs.

declares a function that accepts a "unique_ptr of an `int`s that is `const`"